// @algo-builder/runtime/build/errors/runtime-errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RuntimeError = exports.RUNTIME_ERROR_PREFIX = void 0;
const web_1 = require("@algo-builder/web");
exports.RUNTIME_ERROR_PREFIX = "RUNTIME_ERR";
function getRuntimeErrorCode(error) {
    return `${exports.RUNTIME_ERROR_PREFIX}${error.number}`;
}
class RuntimeError extends Error {
    constructor(errorDescriptor, messageArguments = {}, parentError) {
        const prefix = `${getRuntimeErrorCode(errorDescriptor)}: `;
        const formattedMessage = (0, web_1.applyErrorMessageTemplate)(errorDescriptor.message, messageArguments);
        super(String(prefix) + String(formattedMessage));
        this.errorDescriptor = errorDescriptor;
        this.number = errorDescriptor.number;
        if (parentError instanceof Error) {
            this.parent = parentError;
        }
        this._isRuntimeError = true;
        Object.setPrototypeOf(this, RuntimeError.prototype);
    }
    static isRuntimeError(other) {
        return (other !== undefined && other !== null && other._isRuntimeError === true);
    }
}
exports.RuntimeError = RuntimeError;
