// @algo-builder/algob/build/internal/deployer_cfg.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeployerConfig = exports.mkDeployer = void 0;
const runtime_1 = require("@algo-builder/runtime");
const account_1 = require("../lib/account");
const driver_1 = require("../lib/driver");
const script_checkpoints_1 = require("../lib/script-checkpoints");
const deployer_1 = require("./deployer");
const tx_log_writer_1 = require("./tx-log-writer");
function mkDeployer(allowWrite, deployerCfg) {
    if (allowWrite) {
        return new deployer_1.DeployerDeployMode(deployerCfg);
    }
    return new deployer_1.DeployerRunMode(deployerCfg);
}
exports.mkDeployer = mkDeployer;
// intialize deployer config obj
class DeployerConfig {
    constructor(runtimeEnv, algoOp) {
        this.runtimeEnv = runtimeEnv;
        this.cpData = (0, script_checkpoints_1.loadCheckpointsRecursive)();
        this.algoOp = algoOp;
        this.accounts = (0, account_1.mkAccountIndex)(runtimeEnv.network.config.accounts ?? []);
        this.txWriter = new tx_log_writer_1.TxWriterImpl('');
        this.asaDefs = (0, runtime_1.loadASAFile)(this.accounts);
        this.indexerClient = (0, driver_1.createIndexerClient)(runtimeEnv.network.config.indexerCfg);
    }
}
exports.DeployerConfig = DeployerConfig;
