// @algo-builder/web/build/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mkTxParams = exports.getSuggestedParams = exports.WallectConnectSession = exports.WebMode = exports.ERROR_RANGES = exports.ERRORS = exports.parseAlgorandError = exports.applyErrorMessageTemplate = exports.BuilderError = exports.ASADefsSchema = exports.ASADefSchema = exports.types = exports.tx = exports.status = exports.parsing = void 0;
exports.parsing = __importStar(require("./lib/parsing"));
exports.status = __importStar(require("./lib/status"));
exports.tx = __importStar(require("./lib/txn"));
exports.types = __importStar(require("./types"));
var types_input_1 = require("./types-input");
Object.defineProperty(exports, "ASADefSchema", { enumerable: true, get: function () { return types_input_1.ASADefSchema; } });
Object.defineProperty(exports, "ASADefsSchema", { enumerable: true, get: function () { return types_input_1.ASADefsSchema; } });
var errors_1 = require("./errors/errors");
Object.defineProperty(exports, "BuilderError", { enumerable: true, get: function () { return errors_1.BuilderError; } });
Object.defineProperty(exports, "applyErrorMessageTemplate", { enumerable: true, get: function () { return errors_1.applyErrorMessageTemplate; } });
Object.defineProperty(exports, "parseAlgorandError", { enumerable: true, get: function () { return errors_1.parseAlgorandError; } });
var errors_list_1 = require("./errors/errors-list");
Object.defineProperty(exports, "ERRORS", { enumerable: true, get: function () { return errors_list_1.ERRORS; } });
Object.defineProperty(exports, "ERROR_RANGES", { enumerable: true, get: function () { return errors_list_1.ERROR_RANGES; } });
var web_mode_1 = require("./lib/web-mode");
Object.defineProperty(exports, "WebMode", { enumerable: true, get: function () { return web_mode_1.WebMode; } });
var wallectconnect_mode_1 = require("./lib/wallectconnect-mode");
Object.defineProperty(exports, "WallectConnectSession", { enumerable: true, get: function () { return wallectconnect_mode_1.WallectConnectSession; } });
var api_1 = require("./lib/api");
Object.defineProperty(exports, "getSuggestedParams", { enumerable: true, get: function () { return api_1.getSuggestedParams; } });
Object.defineProperty(exports, "mkTxParams", { enumerable: true, get: function () { return api_1.mkTxParams; } });
