// @algo-builder/web/build/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChainType = exports.isSDKTransactionAndSign = exports.isSDKTransaction = exports.isRequestError = exports.isFileError = exports.TransactionType = exports.SignType = void 0;
var SignType;
(function (SignType) {
    SignType[SignType["SecretKey"] = 0] = "SecretKey";
    SignType[SignType["LogicSignature"] = 1] = "LogicSignature";
})(SignType = exports.SignType || (exports.SignType = {}));
var TransactionType;
(function (TransactionType) {
    TransactionType[TransactionType["TransferAlgo"] = 0] = "TransferAlgo";
    TransactionType[TransactionType["TransferAsset"] = 1] = "TransferAsset";
    TransactionType[TransactionType["ModifyAsset"] = 2] = "ModifyAsset";
    TransactionType[TransactionType["FreezeAsset"] = 3] = "FreezeAsset";
    TransactionType[TransactionType["RevokeAsset"] = 4] = "RevokeAsset";
    TransactionType[TransactionType["DestroyAsset"] = 5] = "DestroyAsset";
    TransactionType[TransactionType["CallApp"] = 6] = "CallApp";
    TransactionType[TransactionType["ClearApp"] = 7] = "ClearApp";
    TransactionType[TransactionType["CloseApp"] = 8] = "CloseApp";
    TransactionType[TransactionType["DeleteApp"] = 9] = "DeleteApp";
    TransactionType[TransactionType["DeployASA"] = 10] = "DeployASA";
    TransactionType[TransactionType["DeployApp"] = 11] = "DeployApp";
    TransactionType[TransactionType["OptInASA"] = 12] = "OptInASA";
    TransactionType[TransactionType["OptInToApp"] = 13] = "OptInToApp";
    TransactionType[TransactionType["UpdateApp"] = 14] = "UpdateApp";
})(TransactionType = exports.TransactionType || (exports.TransactionType = {}));
// This function is used to check if given objects implements `FileError` interface
function isFileError(object) {
    return Object.prototype.hasOwnProperty.call(object, "errno");
}
exports.isFileError = isFileError;
// This function is used to check if given objects implements `RequestError` interface
// https://www.technicalfeeder.com/2021/02/how-to-check-if-a-object-implements-an-interface-in-typescript/
function isRequestError(object) {
    const res = Object.prototype.hasOwnProperty.call(object, "response.statusCode") &&
        Object.prototype.hasOwnProperty.call(object, "response.text") &&
        Object.prototype.hasOwnProperty.call(object, "response.body.message") &&
        Object.prototype.hasOwnProperty.call(object, "response.error");
    return res && Object.prototype.hasOwnProperty.call(object, "error");
}
exports.isRequestError = isRequestError;
// This function checks if given object implements `Transaction` class
function isSDKTransaction(object) {
    if (object === undefined || object === null) {
        return false;
    }
    const props = [
        "tag", "from", "fee", "firstRound", "lastRound",
        "genesisID", "genesisHash"
    ];
    let res = Object.prototype.hasOwnProperty.call(object, "name");
    for (const prop of props) {
        res = res && Object.prototype.hasOwnProperty.call(object, prop);
    }
    return res;
}
exports.isSDKTransaction = isSDKTransaction;
// This function checks if given object implements `Transaction` class and has Sign
function isSDKTransactionAndSign(object) {
    if (object === undefined || object === null) {
        return false;
    }
    const res = isSDKTransaction(object.transaction);
    return Object.prototype.hasOwnProperty.call(object, "sign") && res;
}
exports.isSDKTransactionAndSign = isSDKTransactionAndSign;
/* Wallet Connect types */
var ChainType;
(function (ChainType) {
    ChainType["MainNet"] = "MainNet";
    ChainType["TestNet"] = "TestNet";
})(ChainType = exports.ChainType || (exports.ChainType = {}));
